File: backend/app/exercise_loader.py
"""Exercise configuration loader — reads JSON configs from the exercises/ directory."""

import json
import logging
from pathlib import Path

logger = logging.getLogger("ekusasaizu.exercises")

_EXERCISES_DIR = Path(__file__).parent / "exercises"
_cache: dict[str, dict] = {}


def _load_all() -> dict[str, dict]:
    """Load all exercise configs from JSON files."""
    if _cache:
        return _cache

    if not _EXERCISES_DIR.exists():
        logger.warning("Exercises directory not found: %s", _EXERCISES_DIR)
        return _cache

    for path in sorted(_EXERCISES_DIR.glob("*.json")):
        try:
            with open(path, encoding="utf-8") as f:
                config = json.load(f)
            exercise_id = config.get("id", path.stem)
            _cache[exercise_id] = config
            logger.info("Loaded exercise config: %s (%s)", exercise_id, path.name)
        except (json.JSONDecodeError, KeyError) as e:
            logger.error("Failed to load exercise config %s: %s", path.name, e)

    return _cache


def list_exercises() -> list[dict]:
    """Return a summary list of all available exercises."""
    configs = _load_all()
    return [
        {
            "id": cfg["id"],
            "name": cfg["name"],
            "type": cfg["type"],
            "description": cfg["description"],
            "camera_angle": cfg.get("camera_angle", "front"),
        }
        for cfg in configs.values()
    ]


def get_exercise_config(exercise_id: str) -> dict | None:
    """Return the full MediaPipe configuration for an exercise."""
    configs = _load_all()
    return configs.get(exercise_id)


def reload_exercises() -> None:
    """Force reload all exercise configs from disk."""
    _cache.clear()
    _load_all()

File: backend/src/exercise-loader.ts
/**
 * Exercise configuration loader — reads JSON configs from the exercises/ directory.
 */

import { readdir, readFile } from 'fs/promises';
import { join, basename } from 'path';

const EXERCISES_DIR = join(import.meta.dir, 'exercises');

interface ExerciseSummary {
  id: string;
  name: string;
  type: string;
  description: string;
  camera_angle: string;
}

const cache = new Map<string, Record<string, any>>();

async function loadAll(): Promise<Map<string, Record<string, any>>> {
  if (cache.size > 0) return cache;

  try {
    const files = await readdir(EXERCISES_DIR);
    for (const file of files.sort()) {
      if (!file.endsWith('.json')) continue;
      try {
        const raw = await readFile(join(EXERCISES_DIR, file), 'utf-8');
        const config = JSON.parse(raw);
        const id = config.id ?? basename(file, '.json');
        cache.set(id, config);
        console.log(`Loaded exercise config: ${id} (${file})`);
      } catch (e) {
        console.error(`Failed to load exercise config ${file}:`, e);
      }
    }
  } catch {
    console.warn(`Exercises directory not found: ${EXERCISES_DIR}`);
  }

  return cache;
}

export async function listExercises(): Promise<ExerciseSummary[]> {
  const configs = await loadAll();
  return Array.from(configs.values()).map(cfg => ({
    id: cfg.id,
    name: cfg.name,
    type: cfg.type,
    description: cfg.description,
    camera_angle: cfg.camera_angle ?? 'front',
  }));
}

export async function getExerciseConfig(exerciseId: string): Promise<Record<string, any> | null> {
  const configs = await loadAll();
  return configs.get(exerciseId) ?? null;
}

export function reloadExercises(): void {
  cache.clear();
}
